src/formula.py

```diff
@@ -168,6 +160,42 @@ def install(self) -> None:
                     shutil.copyfileobj(src, dst)
                 output_path.chmod(info.mode)
 
+# something split into e.g. bin/, lib/, include/
+class UnixPackageFormula(Formula):
+    archive_url: str
+
+    @contextlib.contextmanager
+    def _read_archive(self, filename: str, fp: BinaryIO) -> Iterator[Archive]:
+        archive_type = file_archive_type(filename)
+        if archive_type == "zip":
```

Comment:
Maybe pull this into a separate function as well?

Reply:
For context: We read the whole contents of the input file because a file-like from a HTTP response is not seekable, but the file for Zip/Tar archive readers must be seekable.

Reply:
yeah this makes sense. Fwiw in theory it's possible to process tar archives in a streaming fashion, but doesn't seem important here. If memory became an issue you'd just download to a temporary file, use that, then delete.

Reply:
zips are a bit harder because overwrites in a zip file ~= appending a new zip archive with the new file to the end of the zip file.